frame/node-authorization/src/lib.rs

```diff
@@ -0,0 +1,761 @@
+//! It provides two ways to authorize a node,
+//!
+//! - a set of well known nodes across different organizations in which the
+//! connections are allowed.
+//! - users can claim the ownership for each node, then manage the connections of
+//! the node.
+//!
+//! A node must have an owner. The owner can additionally change the connections
+//! for the node. Only one user is allowed to claim a specific node. To eliminate
+//! false claim, the maintainer of the node should claim it before even start the
+//! node. This pallet use offchain work to set reserved nodes, if the node is not
+//! an authority, make sure to enable offchain worker with the right CLI flag.
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use sp_std::prelude::*;
+use sp_core::NodePublicKey;
+use frame_support::{
+    decl_module, decl_storage, decl_event, decl_error,
+    debug, ensure,
+    weights::{DispatchClass, Weight},
+    traits::{Get, EnsureOrigin},
+};
+use frame_system::ensure_signed;
+
+pub trait WeightInfo {
+    fn add_well_known_node() -> Weight;
+    fn remove_well_known_node() -> Weight;
+    fn swap_well_known_node() -> Weight;
+    fn reset_well_known_nodes() -> Weight;
+    fn claim_node() -> Weight;
+    fn remove_claim() -> Weight;
+    fn transfer_node() -> Weight;
+    fn add_connections() -> Weight;
+    fn remove_connections() -> Weight;
+}
+
+impl WeightInfo for () {
+    fn add_well_known_node() -> Weight { 50_000_000 }
+    fn remove_well_known_node() -> Weight { 50_000_000 }
+    fn swap_well_known_node() -> Weight { 50_000_000 }
+    fn reset_well_known_nodes() -> Weight { 50_000_000 }
+    fn claim_node() -> Weight { 50_000_000 }
+    fn remove_claim() -> Weight { 50_000_000 }
+    fn transfer_node() -> Weight { 50_000_000 }
+    fn add_connections() -> Weight { 50_000_000 }
+    fn remove_connections() -> Weight { 50_000_000 }
+}
+
+pub trait Trait: frame_system::Trait {
+    /// The event type of this module.
+	type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+
+    /// The maximum number of well known nodes that are allowed to set
+    type MaxWellKnownNodes: Get<u32>;
+
+    /// The origin which can add a well known node.
+    type AddOrigin: EnsureOrigin<Self::Origin>;
+
+    /// The origin which can remove a well known node.
+    type RemoveOrigin: EnsureOrigin<Self::Origin>;
+
+    /// The origin which can swap the well known nodes.
+    type SwapOrigin: EnsureOrigin<Self::Origin>;
+
+    /// The origin which can reset the well known nodes.
+    type ResetOrigin: EnsureOrigin<Self::Origin>;
+
+    /// Weight information for extrinsics in this pallet.
+    type WeightInfo: WeightInfo;
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as NodeAuthorization {
+        /// The set of well known nodes. This is stored sorted (just by value).
+        pub WellKnownNodes get(fn well_known_nodes): Vec<NodePublicKey>;
```

Comment:
It seems that `BTreeSet` could be a better option instead of `Vec`. We anyway keep:
1. encoding/decoding it 
2. doing `binary_search` (`log n`), but resulting `insert` operation might be linear since it has to move other items (`O(n)`)

`BTreeSet` seems better suited for this workflow, prevents duplicates by design, and we would only pay the cost of converting from `BTreeSet -> Vec` in the offchain worker, which seems acceptable.

Reply:
Same with `AdditionalConnections`.

Reply:
I tried to use BTreeSet, but got compile error in primitive/io offchain runtime interface,
```
the trait `storage::proc_macro_runtime_interface::pass_by::PassBy` is not implemented for `std::collections::BTreeSet<sp_core::NodePublicKey>`
```
I guess it's not supported here, likely a missing feature.
To work around this, I need to convert BTreeSet to Vector in pallet before calling the offchain api. I doubt if it's still worth on this refactor.
Also vector is a bit more performance in error scenarios.

Reply:
You'd either need to add `PassBy` implementation to `BTreeSet`, but I rather think we should use `BTreeSet` on the storage layer, but then convert to `Vec` when passing to `primitive/io` interface.

> I doubt if it's still worth on this refactor.

I disagree. The `binary_search` + `insert/remove` will be simplified to single method call. duplicates are handled automatically in `Set`, so I feel it will simplify the runtime code quite a lot and...

> Also vector is a bit more performance in error scenarios.

...we will probably have small number of elements in the `Vec` anyway so it doesn't really matter that much, but asymptotically speaking `Vec` will never be faster than `BTreeSet` in the operations you do. I suppose you mean that pushing an element to a `Vec` is faster than insertion to the `Set` - true, but in the runtime we should rather care about the worst case (possible attack vector). So again, asymptotically, we always do `O(log n) /* binsearch */ + O(n) /* insert */` in `Vec` case, while `BTreeSet` should only have `O(log n) /* insert/remove */` operations.
That said I believe the main advantage will be code simplification, non-linear cost in non-trivial cases is a nice addition.